Extend the scripts to auto-download the maven-wrapper.jar from maven central

chrisdutz opened the pull request:
I extended the scripts to allow usage of maven-wrapper without requiring the jar to be checked in with the rest of the sources. If the jar is found (because it is checked in or if it had been downloaded peviously) this is used, if not several options are tried out depending on the platform:
- Windows: 
    - powershel is used to download

- Linux/Mac: 
    - curl
    - wget 
    - if both are not available a java class is compiled and executed to do the download.

I didn't add the java option to the windows script for now as far as I know the powershel option should be available on all windows systems.

Commit:
- Added an initial version of an auto-downloader for the maven-wrapper.jar

Commit:
Cleaned up the code a little to not exit with 0 in case of an error.

Commit:
- Made the script use curl or wget if available and fallback to java in case none of them are.
- Moved the Java source code to the filesystem.

Commit:
- Added a download extension for windows

Commit:
- Added Apache header to the downloader java class

atanasenko commented:
Wrapper download url should also be configurable, same as with maven distro that gets downloaded by the wrapper

gzm55 commented on mvnw:
```diff
@@ -200,6 +200,42 @@ if [ -z "$BASE_DIR" ]; then
+    echo "Found .mvn/wrapper/maven-wrapper.jar"
+else
+    echo "Couldn't find .mvn/wrapper/maven-wrapper.jar, downloading it ..."
+    if which wget >/dev/null; then
```
use posix way to check cmd: `command -v wget > /dev/null`

chrisdutz replied:
Ok ... just pushed that change (I really would like to have github send me emails on responses like this)

davidrabinowitz commented:
+1

Commit:
- Changed the curl and wget detection to the "possix way" using "command"

chrisdutz commented:
Right now I'm a little confused as to what is still missing ... if you guys could sum that up for me, I'd be more than happy to address those things. But keep in mind ... I am a total new-be to this writing of sh scripts ... so please write things as if you were explaining things to a child ;-) ("making url configurable" -(does this translate to)-> "add a property to maven-wrapper.properties containing the url and use that in the script")

chrisdutz commented:
I just had a look and it seems as if maven-wrapper.jar uses maven-wrapper.properties to find where to download things ... this isn't used by the mvnw script directly ... should we be mixing things here? If yes ... how could I load the properties in there and make them accessible in the script?

gzm55 commented on mvnw:
```diff
@@ -200,6 +200,42 @@ if [ -z "$BASE_DIR" ]; then
+    echo "Couldn't find .mvn/wrapper/maven-wrapper.jar, downloading it ..."
+    if command -v wget > /dev/null; then
+        echo "Found wget ... using wget"
+        wget ‐‐output-document=.mvn/wrapper/maven-wrapper.jar "$JAR_URL"
```
in some environment, such as in alpine, wget has no ‐‐output-document option, only '-O'  works.

chrisdutz replied:
Does "-O" work everywhere?

gzm55 replied:
almost, at least better than ‐‐output-document

chrisdutz replied:
Ok ... so I updated the pull request.

gzm55 commented:
For a configurable url, we can add a optional key (e.g. wrapperUrl) in maven-wrapper.properties, and parse it manually first before downloading the wrapper jar.

For linux, we can parse properties file like
```sh
jarUrl='a-default-url'
while IFS="=" read key value; do
  case "$key" in
  (wrapperUrl) jarUrl="$value"; break ;;
  esac
done < "path/to/maven-wrapper.properties"
```

For windows, the same logic works, but need to translate to for-statement in cmd script of windows.

Commit:
- Changed the wget command to use "-O" instead of "--output-document="

chrisdutz commented:
So the last change you would like to have me do, would be to make the URL configurable? This Tree-Style-Comments are making it really hard to follow the conversation.

Commit:
- Made the wrapperUrl configurable from the maven-wrapper.properties file
  - Updated and tested the linux/unix/mac version of the script
  - Updated and tested the Java downloader
  - Updated the windows cmd (but not tested yet)

(Need to test the changes to the windows command on a windows machine)

chrisdutz committed:
Fixed the windows version of the script

chrisdutz commented:
Ok ... so now the downloadUrl has a default and is overridable in the properties file ... anything else missing?

gzm55 commented on mvnw:
```diff
@@ -200,6 +200,49 @@ if [ -z "$BASE_DIR" ]; then
+    if command -v wgety > /dev/null; then
+        echo "Found wget ... using wget"
+        wget ‐O "$BASE_DIR/.mvn/wrapper/maven-wrapper.jar" "$jarUrl"
+    elif command -v curly > /dev/null; then
```
`wgety` and `curly` seems typo of `wget` and `curl`, is it?

chrisdutz committed:
Removed a hack to force the script to use curl or wget ...

chrisdutz commented:
Again ... anything else missing?

mosabua commented:
Mostly time to review and test. Before we merge I need to do another feature request (removing output and making it configurable) and this would need to be tested on linux, osx, window and solaris

mosabua commented:
So I cut a new release without this. I am not sure how we can get this tested on the current supported shell and OS versions to be honest.

mosabua commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -1 +1,2 @@
-distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.0/apache-maven-3.5.0-bin.zip
\ No newline at end of file
+distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.0/apache-maven-3.5.0-bin.zip
+#wrapperUrl=http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar
```
Should be the https URL and also be the same repo1.maven.org as above

mosabua replied:
Also should this really be commented out?

chrisdutz replied:
Well it is in there as an example of how to configure the wrapper. It is equal to the default used in the wrapper and downloader code. Sort of wanted to imply that you don't need that property, but you could use it and how you would have to use it.

mosabua commented on .mvn/wrapper/MavenWrapperDownloader.java:
```diff
@@ -0,0 +1,110 @@
+     * Default URL to download the maven-wrapper.jar from, if no 'downloadUrl' is provided.
+     */
+    private static final String DEFAULT_DOWNLOAD_URL =
+            "http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar";
```
Should be HTTPS and repo1.maven.org... and maybe have version set somehow more cleanly.

chrisdutz replied:
I did change the urls to the https versions.

chrisdutz committed:
Updated the download url to the https version

chrisdutz commented:
Ok so I adressed the https issue and just comitted my changes. One way to test this on a variety of systems would be for me to use the updated wrapper in one of the Apache Projects I'm working on. There I could run builds on Windows, Mac, multiple Linux distributions. I did manually test it on Mac, Linux and Windows.

chrisdutz committed:
Merge branch 'master' of github.com:takari/maven-wrapper

chrisdutz committed:
Updated to the latest maven-wrapper version.

chrisdutz commented:
Also just updated the version of the jar to the latest version. It would be easy to inject the current version, if the mvnw and mvnw.cmd etc. would be inside the sources so the maven resource plugin could do the filtering and injecting of properties.

mosabua commented:
Replaced by https://github.com/takari/maven-wrapper/pull/70
